lib/adapters/apply-edit-adapter.ts

```diff
@@ -78,6 +89,90 @@ export default class ApplyEditAdapter {
     return { applied }
   }
 
+  private static async handleResourceOperation(edit: CreateFile | RenameFile | DeleteFile): Promise<void> {
+    if (DeleteFile.is(edit)) {
+      const path = Convert.uriToPath(edit.uri)
+      const stats: boolean | Stats = await fsp.lstat(path).catch(() => false)
+      const ignoreIfNotExists = edit.options?.ignoreIfNotExists || true
+
+      if (!stats) {
+        if (ignoreIfNotExists) {
+          return
+        }
+        throw Error(`Target doesn't exist.`)
+      }
+
+      if (stats.isDirectory()) {
+        if (edit.options?.recursive) {
+          return new Promise((resolve, reject) => {
+            rimraf(path, { glob: false }, (err) => {
+              if (err) {
+                reject(err)
+              }
+              resolve()
+            })
+          })
+        }
+        return fsp.rmdir(path, { recursive: edit.options?.recursive })
+      }
+
+      return fsp.unlink(path)
+    }
+    if (RenameFile.is(edit)) {
+      const oldPath = Convert.uriToPath(edit.oldUri)
+      const newPath = Convert.uriToPath(edit.newUri)
+      const exists = await fsp
+        .access(newPath)
+        .then(() => true)
+        .catch(() => false)
+      const ignoreIfExists = edit.options?.ignoreIfExists
+      const overwrite = edit.options?.overwrite
+
+      if (exists && ignoreIfExists && !overwrite) {
+        return
+      }
+
+      if (exists && !ignoreIfExists && !overwrite) {
+        throw Error(`Target exists.`)
+      }
+
+      return fsp.rename(oldPath, newPath)
+    }
+    if (CreateFile.is(edit)) {
+      const path = Convert.uriToPath(edit.uri)
+      const exists = await fsp
+        .access(path)
+        .then(() => true)
+        .catch(() => false)
+      const ignoreIfExists = edit.options?.ignoreIfExists
+      const overwrite = edit.options?.overwrite
+
+      if (exists && ignoreIfExists && !overwrite) {
+        return
+      }
+
+      return fsp.writeFile(path, "")
+    }
+  }
+
+  private static normalize(workspaceEdit: WorkspaceEdit): void {
```

Comment:
I am not a fan of using static methods. Would you be able to make these free functions? If public method, you can directly export them.

Reply:
What do you mean by free? Could you give an example?

Reply:
For example, for the private methods, you can simply move the code to the bottom of the file.
```ts
function normalize(workspaceEdit: WorkspaceEdit): void {
 // code
}
```

Reply:
If the method is already public, you can keep it for now. We have started to avoid using classes as a namespace, especially for new code.

Reply:
Like this ? https://github.com/atom-community/atom-languageclient/pull/140/commits/a699bd57ac50748a2c490de2bacc58f080423115

Reply:
Exactly! Thanks!